text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,207 @@
+[summary]: #summary
+
+The Fabric Gateway is a new component that will run either as a standalone process, or embedded with the peer, and will implement as much of the high-level 'gateway' programming model as possible.  This will remove much of the transaction submission logic from the client application which will simplify the maintenance of the SDKs.  It will also simplify the administrative overhead of running a Fabric network since the client application will only need to connect to a gateway rather than multiple peers across potentially multiple organisations.
+
+# Motivation
+[motivation]: #motivation
+
+The high-level "gateway" programming model has proved to be very popular with users since its introduction in v1.4 in Node SDK.  Since then, it has been implemented in the Java and Go SDKs.  This has led to a large amount of code that has to be maintained in the SDKs which increases the cost of implementing new feature which have to be replicated to keep the SDKs in step.  It also increases the cost of creating an SDK for a new language.  Moving all of the high-level logic into its own process and exposing a simplified GRPC interface will drastically reduce the complexity of the SDKs.
+
+By providing a single entry point to a Fabric network, client applications can interact with complex network topologies without multiple ports having to be opened and secured to each peer and ordering node.  The client will connect only to the gateway which will act on its behalf with the rest of the network.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The Fabric Gateway is an embodiment of the high-level 'gateway' Fabric programming model in a server component that will form part of a Fabric network alongside Peers, Orderers and CAs.  It will either be stood up inside its own process in a docker container, or it will be hosted inside a peer.  Either way, it exposes its functionality to clients via a GRPC interface.
+
+Lightweight client SDKs are used to connect to the Gateway for the purpose of invoking transactions.  The gateway will intereact with the rest of the network on behalf of the client eliminating the need for the client application to connect directly to peers and orderers.
+
+The concepts and APIs will be familiar to Fabric programmers who are already using the new programming model.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The Gateway runs as a client process associated with an organisation and requires an identity (issued by the org's CA) in order to interact with the discovery service and to register for events (Deliver Client).  
+
+The Gateway also runs as a GRPC server and exposes the following services to client applications (SDKs):
+
+```
+service Gateway {
+    rpc Prepare(SignedProposal) returns (PreparedTransaction) {}
+    rpc Commit(PreparedTransaction) returns (stream Event) {}
+    rpc Evaluate(SignedProposal) returns (Result) {}
```

Comment:
In database/transactional system, prepare means different things. There is a prepared statement and also there is a preparation step after the tx execution but before the commit (2-phase commit in a standalone DB -- prepare(), commit(). Here, the prepare acquires all necessary lock before the commit). 

To avoid this confusion, can't we reflect the fabric architecture in these APIs? One such example is below
```
Execute(SignedProposal) returns (ExecutionResults) {}
Commit(ExecutionResults) returns (stream Event) {}
```
We can also come up with something better but my point is that the prepare does not convey execution.

Reply:
This is a really interesting point. I don't speak for Andy, but I think the intention with this naming is to make it relatable to people coming to Fabric who are more likely to be familiar with concepts like two-phase commit. The prepare gets you ready to be able to successfully commit, which is similar to what you are doing by getting a transaction proposal endorsed.

Looking at it from the other end, in the Fabric programming model exposed by the SDKs we talk about _evaluate_ and _submit_ of transactions. Evaluate is essentially a query where we send a proposal to peer(s) for simulation to get a transaction result; we don't care about endorsement. Submit involves getting peer(s) to endorse the proposal before sending a transaction containing the endorsements to the orderer to be included in a block, and waiting for the block to be committed to peer ledgers. So at the client side I might be thinking of the submit flow consisting of a call to `Endorse` followed by a `Submit` to ordering.

I worry that renaming `Prepare` to `Execute` makes it sound like it is behaving like `Evaluate`; just executing a transaction rather than preparing to commit a transaction by gaining endorsement.

I guess my preference would probably be to either have:
- `Prepare` / `Commit` to imply the sequential relationship between them, as in two-phase commit
- `Endorse` / `Commit` to align with terminology used by Fabric
- `Endorse` / `Submit` to align with the higher level terminology used by the client SDKs

Reply:
In 2 PC, once the prepare succeeds, the commit will be successful. Not sure whether the same semantics apply here. The last two options look good to me.

Reply:
I don't feel too strongly about the names we give to these GRPC functions.  The end users will be using the SDK terminology which is already well established

Reply:
I don't think we should introduce new terminology for these APIs .... 

I also believe that this should be built external to the peer but with thought it mind that it can be embedded in any other process.

Reply:
> I don't think we should introduce new terminology for these APIs ....

I've changed the grpc API to Endorse/Submit

> I also believe that this should be built external to the peer but with thought it mind that it can be embedded in any other process.

Understood, that's not a problem.  The design will remain the same, it just affects the order in which we deliver the user-stories.